journalpump/journalpump.py

```diff
@@ -397,76 +436,203 @@ def send_messages(self, message_batch):
+        }
+
+    def inc_line_stats(self, *, journal_lines, journal_bytes):
+        self.read_bytes += journal_bytes
+        self.read_lines += journal_lines
+
+        now = time.monotonic()
+        if (now - self.last_stats_send_time) < 10.0:
+            # Do not send stats too often
+            return
+
+        tags = self.make_tags()
+        self.stats.gauge("journal.last_read_ago", value=now - self.last_journal_msg_time, tags=tags)
+        self.stats.gauge("journal.read_lines", value=self.read_lines, tags=tags)
+        self.stats.gauge("journal.read_bytes", value=self.read_bytes, tags=tags)
+        self.last_stats_send_time = now
+        lines_diff = self.read_lines - self._last_sent_read_lines
+        bytes_diff = self.read_bytes - self._last_sent_read_bytes
+        if lines_diff or bytes_diff:
+            self.log.info("Processed %r journal lines (%r bytes)", lines_diff, bytes_diff)
+        self._last_sent_read_lines = self.read_lines
+        self._last_sent_read_bytes = self.read_bytes
+
+        for sender in self.senders.values():
+            sender.refresh_stats()
+
+    def read_next(self):
+        journald_reader = self.get_reader(seek_to=self.cursor)
+        if journald_reader:
+            jobject = next(journald_reader)
+            if jobject.cursor:
+                self.cursor = jobject.cursor
+                self.last_journal_msg_time = time.monotonic()
+        else:
+            jobject = None
+
+        inactivity_timeout = 180
+        if (time.monotonic() - self.last_journal_msg_time) > inactivity_timeout and self.cursor:
+            self.log.info("We haven't seen any msgs in %.1fs, reinitiate PumpReader()", inactivity_timeout)
+            self.get_reader(seek_to=self.get_resume_cursor(), reinit=True)
+            self.last_journal_msg_time = time.monotonic()
+
+        return jobject
+
+    def get_reader(self, seek_to=None, reinit=False):
+        """Return an initialized reader or None"""
+        if not reinit and self.journald_reader:
+            return self.journald_reader
 
-    def init_reader(self):
         if self.journald_reader:
+            # Close the existing reader
             self.journald_reader.close()  # pylint: disable=no-member
             self.journald_reader = None
 
-        if self.config.get("journal_path"):
-            while self.running:
-                try:
-                    self.journald_reader = PumpReader(path=self.config["journal_path"])
-                    break
-                except FileNotFoundError as ex:
-                    self.log.warning("journal not available yet, waiting: %s: %s",
-                                     ex.__class__.__name__, ex)
-                    time.sleep(5.0)
-        else:
-            self.journald_reader = PumpReader()
+        try:
+            self.journald_reader = PumpReader(
+                files=self.config.get("journal_files"),
+                flags=self.config.get("journal_flags"),
```

Comment:
I'm not sure what's the right place to do the conversion, but I'd like to see something like this:
```
flags = self.config.get("journal_flags")
if isinstance(flags, str):
    flags = reduce(
        lambda a, b: a | b,
        [getattr(systemd.journal, flag.strip()) for flag in flags.split("|")],
    )
```
so users wouldn't have to write integer flags in their configuration files.